### ISSUE TITLE
gnURL is being build with the wrong configure switches

### ISSUE THREAD

**Issue by ng-0** [NixOS/nixpkgs#30219]
I find the configure flags passed to gnurl in Nix curious.

Here is what I use in Guix, and how it is intended to be build:

https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/gnunet.scm?id=6a5f7af6fccb7681b5e1a9fbb1c772033081f8d3#n186

to paste the relevant parts:

```
    `(#:configure-flags '("--enable-ipv6" "--with-gnutls" "--without-libssh2"
                          "--without-libmetalink" "--without-winidn"
                          "--without-librtmp" "--without-nghttp2"
                          "--without-nss" "--without-cyassl"
                          "--without-polarssl" "--without-ssl"
                          "--without-winssl" "--without-darwinssl"
                          "--disable-sspi" "--disable-ntlm-wb"
                          "--disable-ldap" "--disable-rtsp" "--disable-dict"
                          "--disable-telnet" "--disable-tftp" "--disable-pop3"
                          "--disable-imap" "--disable-smtp" "--disable-gopher"
                          "--disable-file" "--disable-ftp" "--disable-smb")
```
You however pass:

```
  configureFlags = [
    "--with-zlib"
    "--with-gssapi"
    "--with-libssh2"
    "--with-libidn"
    "--with-libpsl"
    "--enable-ldap"
    "--enable-ldaps"
];
```

Which from my point of view results in a limited cURL, not gnURL as described on https://gnunet.org/gnurl

I'm considering to add an --enable-gnurl switch to configure (or try out the new "build just http" in combination with some other siwtches) so that I don't have to explicitly tell people how gnURL is supposed to be build.

As I don't contribute to Nix nor plan to, someone should fix this.

**Comment by knedlsepp** [comment]
What are the implications of using the flags you are suggesting? There are a lot of `--disable-*` there. It doesn't seem that we gain any new functionality from adding a disable flag.

**Comment by ng-0** [comment]
I would suggest to read https://gnunet.org/gnurl for the answer to that question.

The point in this issue is that you seem to leave out what makes gnURL gnURL, a set of pre-defined configurations and a limited subset of cURL (basically just HTTP+HTTPS build against GnuTLS as SSL backend).

**Comment by knedlsepp** [comment]
Ok. After reading the link you provided I get that the crypto backends should be disabled to decrease the attack surface.
However I am surprised that `gnurl` even made it into nixpkgs, considering that the technical issues for this fork of curl to exist are basically nonexistant on NixOS.
As far as I understand every person worrying about system security could just override the `curl` expression and disable those services.
Additionally it seems a little odd that a library meant to only support http/https and gnutls must explicitly be told during configuration that it shouldn't enable other backends. Maybe we should submit a bug report upstream?

**Comment by ng-0** [comment]
I'm the upstream maintainer (and part of the upstream project, GNUnet).
Let's say "it is complicated".
The short story is we don't want GNUnet to depend on the whole kitchen when we just require the knife, fork and an oven. gnURL started out as a minimal fork of cURL that you still need to build and configure somehow with the cURL build-system.
After a bit of work in the last versions, since late 2016(?) when I took over maintainership of gnURL I met at the original goals of it.
Now we don't want this, but it goes deep down in cURL code to get us what we would want without gnURL. You can search our bugtracker (https://gnunet.org/bugs/search.php?project_id=13&search=gnurl&sticky=on&sort=last_updated&dir=DESC&hide_status=90&match_type=0) for recent discussions on this.
The underlying cURL issues are on cURLs roadmap, but it is no task that can be done quickly.

We are investigating libwget2 to get rid of gnURL, but the need for gnURL will stick around for quiet some time until (if at all) something smaller is found.

**Comment by knedlsepp** [comment]
@falsifian @vrthra Ping.

**Comment by falsifian** [comment]
I can confirm I added `libgnurl` just to support gnunet (c372dc74c2bd583d7b05996cb5a9fc8fc8af19bf).

@ng-0, is it reasonable to expect nothing other than gnunet should ever depend on it? If so, I suggest we remove it from `all-packages.nix`, make it a local variable visible only to the gnunet expression, and make the changes you suggested in your original post.

Or can curl be used as a drop-in replacement? If so I suggest we just delete the `libgnurl` expression. Avoiding the kitchen sink is nice, but probably not worth the extra complexity in this case for the nix project.

**Comment by ng-0** [comment]
No, it is a general purpose library like libcurl. Any other project can use it. So it is no "should not", I'm just not aware of any other software relying on it *at this point*.

What extra complexcity? In the end it is up to you how you resolve this. It is just our *recommendation* in GNUnet to use libgnurl instead of libcurl *or* a curl build exactly like this. Please consult the GNUnet documentation to figure it out for your system (https://gnunet.org/git/gnunet.git/tree/doc or the drupal one with less structure at https://gnunet.org). In Guix we keep it as a public variable available like curl.

**Comment by falsifian** [comment]
The extra complexity is having one more expression in nixpkgs. Additionally I don't think we'd notice if it were broken in some subtle way, as long as gnunet kept working, so there's a risk of having a package waiting to surprise the next person who tries to use it.

Since @knedlsepp is claiming that we can achieve the same effect just by configuring libcurl differently, I'm tempted to just remove libgnurl. But I do not feel strongly about it, and also I am not currently using gnunet, so it's hard for me to test changes.

**Comment by ng-0** [comment]
Well it runs the same testsuite as curl, and I inspect the builds on hydra.gnu.org to check if anything unexpected happens, but it's your call in the end. It is tested.

**Comment by knedlsepp** [comment]
@falsifian: I also think that removing gnurl is probably the way to go (assuming the gnunet expression can be fixed).

**Comment by falsifian** [comment]
@ng-0, thanks for raising this issue and the detailed explanation. To get the ball rolling, I set out: https://github.com/NixOS/nixpkgs/pull/30233. I think the only thing missing is to test it (unfortunately I don't even have time today to verify it builds).

As a side note, I'm curious: how do you inspect the builds? Do you use some sort of automation? It gives me a lot of confidence to hear someone is doing that!

**Comment by falsifian** [comment]
@knedlsepp sorry, I didn't see your message before sending mine.

I'm also willing to follow your suggestion, if you can figure out how to update gnunet. But since @ng-0 has provided easy-to-follow instructions I figured it's simplest just to follow them.

Also, since @ng-0 has more context than me and seems enthusiastic for libgnurl to stay, I am now leaning toward keeping it, even though I don't currently understand how libgnurl is different from libcurl + different config flags.

**Comment by ng-0** [comment]
It's pretty much standard:

curl has a test suite. Since gnURL is just cURL + namespace collisions removed and some more fixes I can't remember right now (see git log) and Guix builds with tests unless you explicitly tell Guix to not run them.
Currently I build on x86_64 and once all tests pass (which is usually always the case unless they also fail in upstream, curl, release) I tag the commit, run my version of maketgz $version-number, and publish the tarball + signature file.
As gnurl is not yet integrated with the CI of gnunet.org my only way to check out other architectures is to either get input from those with the architectures to report bugs and send patches (hint, hint ;)) and to monitor hydra.gnu.org which is the current CI of Guix.
Right after the release I send the version bump for QA to Guix, which is then getting build on hydra.gnu.org after the patch is merged.
To some extent this is here https://git.krosos.org/gnurl-release-spec/tree/ReleaseProcess and https://git.krosos.org/scripts/tree/sh/make_gnurl_release.sh here displayed.

I'd like to get more automation in there (obviously). We'll soon have the CI situation at gnunet.org figured out (there is a CI at the moment, but we have plans to improve it), and then I'd like to make use of it for gnURL aswell, for a multitude of architectures and systems.

**Comment by ng-0** [comment]
The input you provided yesterday lead me start thinking about cutting down on the switches, one of the next versions of gnURL should be more like what Christian had in mind back then, or at least more approachable. Follow either https://gnunet.org/bugs/view.php?id=5143 or the announcement posts on the frontpage of gnunet.org for when you have to change the switches, in case you decide to keep the package.
Thanks for giving me some input on the impression about the configure phase (as a contributor to Guix and other systems this is bothering me aswell)

**Comment by falsifian** [comment]
Fixed by #30233.
